File: src/commands/switch-class.ts
import { CommandInteraction, GuildMember } from "discord.js";
import Command from "../structures/Command.js";
import { groups } from "../utils/Constants.js";

const choices = Object.entries(groups)
	.map(([value, { name }]) => ({ name, value }));

const globalRoles = [
	process.env.ROLE_ID_STUDENT,
	process.env.ROLE_ID_TEACHER,
	process.env.ROLE_ID_ALUMNI,
	process.env.ROLE_ID_GUEST,
];

export default class SwitchclassCommand extends Command {
	constructor() {
		super("changer-classe", {
			description: "Nouvelle année, nouveau groupe !",
			args: {
				groupe: { as: "choice", desc: "Votre classe (ou catégorie)", choices },
			}
		})
	}

	async run(interaction: CommandInteraction) {
		const group = interaction.options.get("groupe").value as string;

		const member = interaction.member as GuildMember;
		
		if (member.roles.cache.find(r => r.name.includes('[2023] '))) {
			interaction.reply({
				content: "Il semblerait que vous ayez déjà sélectionné votre groupe pour l'année universitaire 2023-2024.\nSi vous avez fait une erreur dans la sélection de votre classe, veuillez le signaler aux administrateurs du serveur.",
				ephemeral: true,
			});
		} else {
			const roles = groups[group].roles;

			// warning: bad workaround for stupid issue
			// https://github.com/discordjs/discord.js/issues/8165
			
			// Add roles matching the selected group and remove previously given roles no longer matching the active group
			member.roles.add(roles).then(() => {
				member.roles.remove(globalRoles.filter(r => !roles.includes(r))).then(() => {
					interaction.reply({
						content: "Votre nouveau groupe pour la nouvelle année universitaire a bien été sélectionné !\nDans le doute, vérifiez que vos rôles sont bien en adéquation avec votre groupe.",
						ephemeral: true
					});
				}).catch(err => {
					console.error(err);
					interaction.reply({
						content: "Quelque chose s'est mal passé ! Veuillez contacter les administrateurs du serveur.",
						ephemeral: true
					});
				});
			}).catch(err => {
				console.error(err);
				interaction.reply({
					content: "Quelque chose s'est mal passé ! Veuillez contacter les administrateurs du serveur.",
					ephemeral: true
				});
			});
		}
	}
}

File: src/commands/switch_class.rs
use poise::serenity_prelude::CacheHttp;

use crate::{
    config::{Group, CONFIG},
    Context, Error,
};

/// Nouvelle année, nouveau groupe !
#[poise::command(slash_command, rename = "changer-classe")]
pub async fn changer_classe(
    ctx: Context<'_>,
    #[description = "Votre classe (ou catégorie)"] groupe: Group,
) -> Result<(), Error> {
    let mut member = ctx.author_member().await.unwrap();

    if member.roles(ctx).unwrap().into_iter().any(|role| {
        role.name
            .contains(format!("[{}] ", CONFIG.current_year).as_str())
    }) {
        ctx.send(|reply| {
            reply
                .content(format!(
                    "Il semblerait que vous ayez déjà sélectionné votre groupe pour l'année universitaire {}-{}.\nSi vous avez fait une erreur dans la sélection de votre classe, veuillez le signaler aux administrateurs du serveur.",
                    CONFIG.current_year,
                    CONFIG.current_year.parse::<usize>().unwrap() + 1)
                )
                .ephemeral(true)
        }).await?;

        return Ok(());
    }

    // Add roles matching the selected group and remove previously given roles no longer matching the active group

    member
        .to_mut()
        .remove_roles(ctx.http(), groupe.to_roleids().as_slice())
        .await?;

    member
        .to_mut()
        .add_roles(ctx.http(), groupe.to_roleids().as_slice())
        .await?;

    ctx.send(|reply| {
        reply
            .content("Votre nouveau groupe pour la nouvelle année universitaire a bien été sélectionné !\nDans le doute, vérifiez que vos rôles sont bien en adéquation avec votre groupe.")
            .ephemeral(true)
    }).await?;

    Ok(())
}
